### PR TITLE
Attempt to fix https://github.com/gradle/gradle-private/issues/975

### PR DESCRIPTION
**Pull request by blindpirate**
This PR adds a guard for PipedInputStream to make sure it's safe to
read after being closed.

### Context

See https://github.com/gradle/gradle-private/issues/975

The detail of this issue is still unclear. However, looks like this fixes the issue according to 50~60  AdHoc Functional Tests.

### CODE REVIEW

File: subprojects/core/src/integTest/groovy/org/gradle/api/internal/tasks/userinput/DefaultBuildScanUserInputHandlerIntegrationTest.groovy [hunk 1]
```diff
@@ -16,113 +16,57 @@
...
-        withConsoleOutput(consoleOutput)
+        withDaemon(daemon)
+        withRichConsole(richConsole)
 
         when:
         def gradleHandle = executer.withTasks(DUMMY_TASK_NAME).start()
 
         then:
-        writeToStdIn(gradleHandle, YES.bytes)
+        writeToStdInAndClose(gradleHandle, YES.bytes)
```
Comment (oehme) [inline subprojects/core/src/integTest/groovy/org/gradle/api/internal/tasks/userinput/DefaultBuildScanUserInputHandlerIntegrationTest.groovy]: ❌ Why are we closing stdIn? Maybe that's the whole problem? Shouldn't this be done as part of the teardown process automatically?

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/server/exec/ForwardClientInput.java [hunk 2]
```diff
@@ -84,4 +84,37 @@ public Void call() {
             throw UncheckedException.throwAsUncheckedException(e);
         }
     }
-}
\ No newline at end of file
+
+    private static class CloseAwareInputStream extends PipedInputStream {
```
Comment (oehme) [inline subprojects/launcher/src/main/java/org/gradle/launcher/daemon/server/exec/ForwardClientInput.java]: ⭕️ "close aware" doesn't really say what this does on close. Maybe call it something like "EndOnCloseInputStream"

File: subprojects/core/src/integTest/groovy/org/gradle/api/internal/tasks/userinput/DefaultBuildScanUserInputHandlerIntegrationTest.groovy (hunk 1 above)
Comment (blindpirate) [reply subprojects/core/src/integTest/groovy/org/gradle/api/internal/tasks/userinput/DefaultBuildScanUserInputHandlerIntegrationTest.groovy]: That's exactly my intuition at the beginning: https://github.com/gradle/gradle-private/issues/975#issuecomment-336710468 . This was proven to be wrong after [Ben's change](https://github.com/gradle/gradle/pull/3222). After this change, we met another `IOException: Write end dead` in addition to the original `Pipe closed`. According to [this post](https://techtavern.wordpress.com/2008/07/16/whats-this-ioexception-write-end-dead/), we should close the write end in time.

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/server/exec/ForwardClientInput.java (hunk 2 above)
Comment (blindpirate) [reply subprojects/launcher/src/main/java/org/gradle/launcher/daemon/server/exec/ForwardClientInput.java]: Done.

**Comment by blindpirate** [comment]
I agree on that we need to further investigate it. I'll do it.

**Comment by blindpirate** [comment]
I've created https://github.com/gradle/gradle/issues/3534 as a follow-up issue. Let's merge this as a temporary fix.

**Pull request merged**